R/checkModels.R

```diff
@@ -1,74 +1,170 @@
+      stop(paste0("Column '",x,"' of 'person.groups' is constant."))}
+    if(any(is.na(d[,x]))){
+      stop(paste0("Column '",x,"' of 'person.groups' has missing values."))}
+  })
+  return(d)
+}
+
+### called by defineModel() and splitModels() ----------------------------------
+
+checkQmatrixConsistency <- function(qmat){
+  qmat <- eatTools::makeDataFrame(qmat, name = "Q matrix")
+  if(!inherits(qmat[,1], "character")){
+    qmat[,1] <- as.character(qmat[,1])}
+
+  nClass<- sapply(qmat[,-1,drop=FALSE], inherits, what=c("numeric", "integer"))
+  if(!all(nClass)){
+    warning(paste0("Found non-numeric indicator column(s) in the Q matrix. Transform column(s) '",
+                   paste(colnames(qmat)[ which(nClass==FALSE)+1], collapse = "', '") ,"' into numeric format."))
+    qmat <- data.frame(qmat[,1,drop=FALSE], eatTools::asNumericIfPossible(qmat[,-1,drop=FALSE]), stringsAsFactors = FALSE)}
+
+  werte <- eatTools::tableUnlist(qmat[,-1,drop=FALSE], useNA="always")
+  if(length(setdiff(names(werte), c("0","1", "NA")))<0){
+    stop("Q matrix must not contain entries except '0' and '1'.\n")}
+  if(werte[match("NA", names(werte))] > 0){
+    stop("Missing values in Q matrix.\n")}
+
+  wertes <- lapply(qmat[,-1,drop=FALSE], FUN = function(col) {all(col == 0)})
+  konst <- which(wertes == TRUE)
+  if(length(konst)>0){
+    cat(paste0("Column(s) '",paste(names(konst), collapse = "', '"),"' in Q matrix are konstant with value 0. Delete column(s).\n"))
+    qmat <- qmat[,-match(names(konst), colnames(qmat)), drop=FALSE]
+  }
+
+  doppel <- which(duplicated(qmat[,1]))
+  if(length(doppel)>0){
+    cat("Found duplicated elements in the item id column of the q matrix. Duplicated elements will be removed.\n")
+    chk  <- table(qmat[,1])
+    chk  <- chk[which(chk > 1)]
+    chkL <- lapply(names(chk), FUN = function(ch){
+      qChk <- qmat[which(qmat[,1] == ch),]
+      pste <- apply(qChk, 1, FUN = function(x) {paste(x[-1], collapse="")})
+      if(!all(pste == pste[1])){
+        stop("Inconsistent q matrix.\n")}
+    })
+    qmat <- qmat[!duplicated(qmat[,1]),]
+  }
+
+  zeilen <- apply(qmat, 1, FUN = function(y) {all(names(table(y[-1])) == "0")} )
+  weg    <- which(zeilen == TRUE)
+  if(length(weg)>0) {
+    cat(paste("Note: Following ",length(weg)," item(s) in Q matrix do not belong to any dimension. Delete these item(s) from Q matrix.\n",sep=""))
+    cat("    "); cat(paste(qmat[weg,1],collapse=", ")); cat("\n")
+    qmat <- qmat[-weg,]
+  }
+  return(qmat)
+}
+
 ### called by defineModel() ----------------------------------------------------
 
-checkContextVars <- function(x, varname, type = c("weight", "DIF", "group", "HG"), itemdata, suppressAbort = FALSE, internal = FALSE)   {
+checkContextVars <- function(x, varname, type = c("weight", "DIF", "group", "HG"), itemdata, suppressAbort = FALSE, internal = FALSE){
+  checkmate::assert_character(varname, len = 1)
+  lapply(c(suppressAbort, internal), checkmate::assert_logical, len = 1)
```

Comment:
I mean these. Should i add more? `x` is already checked I think

Reply:
I think no assertations are necessary here. `checkContextVars()` is no user-level function. The input `checkContextVars()` receives has already been "prepared" by `eatTools::existsBackgroundVariables` in an earlier step.